nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayTryFinallyCases.java

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (c) 2017 Uber Technologies, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.uber.nullaway.testdata;
+
+import java.io.IOException;
+import javax.annotation.Nullable;
+
+// This code tests our, admitedly quite unsound, handling of try{ ... }finally{ ... } blocks.
+public class NullAwayTryFinallyCases {
+
+  public void tryFinallyMininal(@Nullable Object o) {
+    try {
+    } finally {
+      // BUG: Diagnostic contains: dereferenced expression
+      System.out.println(o.toString());
+    }
+  }
+
+  public void tryFinallyMininalRet(@Nullable Object o) {
+    try {
+      return;
+    } finally {
+      // This should be an error, but isn't.
+      System.out.println(o.toString());
+    }
+  }
+
+  public void tryFinallyMininalThrow(@Nullable Object o) {
+    try {
+      throw new Error();
+    } finally {
+      // BUG: Diagnostic contains: dereferenced expression
+      System.out.println(o.toString());
+    }
+  }
+
+  public void tryFinallyThrowWFix(@Nullable Object o) {
+    try {
+      o = new Object();
+      throw new Error();
+    } finally {
+      /// ToDo: This should be safe
+      // BUG: Diagnostic contains: dereferenced expression
```

Comment:
I think technically `new Object()` can throw an OOM in which case this could fail

Reply:
True. But then that's the case for every single field initialization. So we should hope OOMs are not modeled. If they are, this could be the underlying reason for the problems with `try{...}finally{...}` and initialization. Basically, then NullAway is right that anything wrapped in the `try` block might be skipped over at runtime, as long as it includes memory allocation (maybe even if not, what happens with an `InterruptedException`?).

Reply:
Maybe the right behavior is to ignore the possibility of unchecked exceptions and handle only checked exceptions?

Reply:
Well I don't think we want to go too nuts with special-case handling of try-catch.  It hasn't really bitten us much yet, I think it's a rare case.  Having the tests are good, though, at least as some kind of documentation for how the tool behaves.  And if it looks like something is buggy upstream (like missing finally after a `return` statement) we should report it.

Reply:
Sounds good to me. Making the comments a bit more clear on this possibility, and will land after that. Will build a test-case for upstream on the other one, and then start hacking in a way to handle the initialization issue without breaking the current behavior too much.